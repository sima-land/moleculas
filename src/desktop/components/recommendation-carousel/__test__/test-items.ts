import { CarouselItem } from '../types';

export const items: CarouselItem[] = [
  {
    product: {
      name: 'Ножницы портновские, 19 см, цвет чёрный',
      imageSrc: 'https://cdn3.static1-sima-land.com/items/3572134/0/280.jpg?v=1594121833',
      url: '',
      price: 261,
      currencyGrapheme: '₽',
      oldPrice: undefined,
    },
    cart: {
      stepText: 'по 20 шт',
      markupText: 'Комплектация + 50 ₽ при покупке до 20 шт',
    },
  },
  {
    product: {
      name: 'Ножницы портновские, 31 см, цвет чёрный/голубой',
      imageSrc: 'https://cdn3.static1-sima-land.com/items/2300382/0/280.jpg?v=1594121798',
      url: '',
      price: 322,
      currencyGrapheme: '₽',
      oldPrice: undefined,
    },
    cart: {
      stepText: 'по 20 шт',
      markupText: 'Комплектация + 50 ₽ при покупке до 20 шт',
    },
  },
  {
    product: {
      name: 'Ножницы портновские, 21,5 см, цвет чёрный/голубой',
      imageSrc: 'https://cdn3.static1-sima-land.com/items/2300383/0/280.jpg?v=1594121773',
      url: '',
      price: 238,
      currencyGrapheme: '₽',
      oldPrice: undefined,
      badges: [
        { color: '#b52ea8', fields: [{ type: 'text', value: '-56%' }] },
        { color: '#ff7200', fields: [{ type: 'text', value: '3 по цене 2' }] },
        { color: '#2962ff', fields: [{ type: 'text', value: 'Товар месяца' }] },
        { color: '#00b8d4', fields: [{ type: 'text', value: 'Акция' }] },
      ],
    },
    cart: {},
  },
  {
    product: {
      name: 'Ножницы портновские, 23 см, цвет МИКС',
      imageSrc: 'https://cdn3.static1-sima-land.com/items/1550239/0/280.jpg?v=1600179085',
      url: '',
      price: 171,
      currencyGrapheme: '₽',
      oldPrice: undefined,
    },
    cart: {
      stepText: 'по 5 шт',
      markupText: 'Комплектация + 50 ₽ при покупке до 20 шт',
      isFetching: true,
    },
  },
  {
    product: {
      name: 'Ножницы портновские, одинаковые кольца, 23 см, цвет чёрный',
      imageSrc: 'https://cdn3.static1-sima-land.com/items/4002852/0/280.jpg?v=1594121894',
      url: '',
      price: 483,
      currencyGrapheme: '₽',
      oldPrice: undefined,
    },
    cart: {},
  },
  {
    product: {
      name: 'Ножницы портновские, 22,5 см, цвет чёрный',
      imageSrc: 'https://cdn3.static1-sima-land.com/items/4005918/0/280.jpg?v=1594121898',
      url: '',
      price: 988,
      currencyGrapheme: '₽',
      oldPrice: undefined,
    },
    cart: {
      stepText: 'по 10 шт',
      markupText: 'Комплектация + 50 ₽ при покупке до 20 шт',
    },
  },
  {
    product: {
      name: 'Ножницы портновские, 22 см, цвет серебряный',
      imageSrc: 'https://cdn3.static1-sima-land.com/items/2300389/0/280.jpg?v=1594121830',
      url: '',
      price: 619,
      currencyGrapheme: '₽',
      oldPrice: undefined,
    },
    cart: {
      stepText: 'по 1 шт',
      qty: 23,
      inCart: true,
    },
  },
  {
    product: {
      name: 'Ножницы портновские, с увеличенным углом наклона, 27 см, цвет чёрный',
      imageSrc: 'https://cdn3.static1-sima-land.com/items/2117838/0/280.jpg?v=1594121766',
      url: '',
      price: 539,
      currencyGrapheme: '₽',
      oldPrice: 595,
    },
    cart: {},
  },
  {
    product: {
      name: 'Ножницы портновские, с прорезиненной ручкой, 20 см, цвет чёрный/серый',
      imageSrc: 'https://cdn3.static1-sima-land.com/items/1108673/0/280.jpg?v=1594103956',
      url: '',
      price: 599,
      currencyGrapheme: '₽',
      oldPrice: 717,
    },
    cart: {},
  },
  {
    product: {
      name: 'Ножницы портновские, скошенное лезвие, 15 см, цвет серебряный',
      imageSrc: 'https://cdn3.static1-sima-land.com/items/4150936/0/280.jpg?v=1580193932',
      url: '',
      price: 253,
      currencyGrapheme: '₽',
      oldPrice: undefined,
    },
    cart: {},
  },
];
