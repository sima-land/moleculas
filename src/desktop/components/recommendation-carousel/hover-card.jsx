import { getOriginCorrection } from '@dev-dep/ui-nucleons/with-tooltip/utils';
import React, { useRef, useState } from 'react';
import { ProductCard } from './product-card';
import classnames from 'classnames/bind';
import styles from './hover-card.scss';
import { omit, pick } from 'lodash';

const cx = classnames.bind(styles);

/**
 * Карточка товара, появляющаяся над элементом карусели.
 * @param {Object} props Свойства компонента.
 * @return {ReactElement} Элемент.
 */
export const HoverCard = ({
  controlRef,
  onAdd,
  onChange,
  onFavoriteClick,
  onQuickViewClick,
  onSubtract,
  ...restProps
}) => {
  const ref = useRef();
  const [itemInfo, setItemInfo] = useState();

  // eslint-disable-next-line require-jsdoc
  const show = (info, target) => {
    setItemInfo(info);

    const element = ref.current;
    const rect = target.getBoundingClientRect();
    const correction = getOriginCorrection(ref.current);

    element.classList.remove(cx('hidden'));
    element.style.left = `${correction.x + rect.left - 16}px`;
    element.style.top = `${correction.y + rect.top - 16}px`;
    element.style.width = `${rect.width + 32}px`;
  };

  // eslint-disable-next-line require-jsdoc
  const hide = () => {
    // не нужно сбрасывать данные тут (например setInfo(null)) так как blur не будет работать из-за размонтирования
    ref.current && ref.current.classList.add(cx('hidden'));
  };

  // eslint-disable-next-line require-jsdoc
  const bindInfo = fn => () => fn && fn(itemInfo);

  // отдаем управление чтобы не поднимать состояние выше
  controlRef.current = controlRef.current || {};
  Object.assign(controlRef.current, { show, hide });

  const [cartData, productData] = [pick, omit].map(fn => fn(itemInfo, [
    'allowFloat',
    'canAdd',
    'canSubtract',
    'inCart',
    'isFetching',
    'markupText',
    'qty',
    'stepText',
  ]));

  return (
    <ProductCard
      {...restProps}
      ref={ref}
      className={cx('card', 'hidden')}
      onMouseLeave={hide}
      inCartControl={{
        ...cartData,
        onAdd: bindInfo(onAdd),
        onChange: newQty => onChange && onChange(itemInfo, newQty),
        onSubtract: bindInfo(onSubtract),
      }}
      productInfo={{
        ...productData,
        onFavoriteClick: bindInfo(onFavoriteClick),
        onQuickViewClick: bindInfo(onQuickViewClick),
      }}
    />
  );
};
