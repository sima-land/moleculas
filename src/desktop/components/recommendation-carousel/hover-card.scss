.card {
  position: absolute;
  opacity: 1;
  visibility: visible;
  &.hidden {
    opacity: 0;
    visibility: hidden;
    pointer-events: none;
  }
}
