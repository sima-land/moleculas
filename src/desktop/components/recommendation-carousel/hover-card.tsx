import React, { useLayoutEffect, useRef, useState } from 'react';
import { getOriginCorrection } from '@sima-land/ui-nucleons/with-tooltip/utils';
import { ProductCard, ProductCardProps } from './product-card';
import classnames from 'classnames/bind';
import styles from './hover-card.module.scss';
import { CarouselItem } from './types';

export interface HoverCardProps extends Omit<ProductCardProps, 'onChange' | 'inCartControl' | 'productInfo'> {
  info: CarouselItem
  targetRef: React.RefObject<HTMLElement | null>
  onAdd?: (info: CarouselItem) => void
  onChange?: (info: CarouselItem, newQty: number) => void
  onMouseLeave?: () => void
  onQuickViewClick?: (info: CarouselItem) => void
  onSubtract?: (info: CarouselItem) => void
}

const cx = classnames.bind(styles);

/**
 * Карточка товара, появляющаяся над элементом карусели.
 * @param props Свойства компонента.
 * @return Элемент.
 */
export const HoverCard = ({
  info,
  targetRef,
  onAdd,
  onChange,
  onMouseLeave,
  onQuickViewClick,
  onSubtract,
  ...restProps
}: HoverCardProps) => {
  const ref = useRef<HTMLDivElement | null>(null);
  const [ready, setReady] = useState<boolean>(false);

  useLayoutEffect(() => {
    if (info && targetRef.current && ref.current) {
      const element = ref.current;
      const rect = targetRef.current.getBoundingClientRect();
      const correction = getOriginCorrection(ref.current);

      element.style.left = `${correction.x + rect.left - 16}px`;
      element.style.top = `${correction.y + rect.top - 16}px`;
      element.style.width = `${rect.width + 32}px`;

      setReady(true);
    }
  }, [info]);

  // eslint-disable-next-line require-jsdoc
  const bindInfo = (fn: any) => () => fn && fn(info);

  return (
    <ProductCard
      {...restProps}
      ref={ref}
      className={cx('card', !ready && 'hidden')}
      onMouseLeave={() => {
        setReady(false);
        onMouseLeave && onMouseLeave();
      }}
      inCartControl={{
        ...info.cart,
        onAdd: bindInfo(onAdd),
        onChange: (newQty: any) => onChange && onChange(info, newQty),
        onSubtract: bindInfo(onSubtract),
      } as any}
      productInfo={{
        ...info.product,
        onQuickViewClick: bindInfo(onQuickViewClick),
      } as any}
    />
  );
};
