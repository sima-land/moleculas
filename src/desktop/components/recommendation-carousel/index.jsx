import React, { useRef, useEffect } from 'react';
import Carousel from '@dev-dep/ui-nucleons/carousel';
import RecommendedItem from './recommended-item';
import CarouselButton from './recommendation-carousel-button';
import withInViewportObserver from '@dev-dep/ui-nucleons/with-in-viewport-observer';
import isFunction from 'lodash/isFunction';
import { isNonEmptyArray } from '../../../common/helpers/is-non-empty';
import classes from './recommendation-carousel.scss';
import classnames from 'classnames/bind';
import Types from 'prop-types';

const cx = classnames.bind(classes);

/**
 * Компонент карусели рекомендованных товаров.
 * @param {Object} props Свойства компонента.
 * @param {string} props.title Заголовок карусели.
 * @param {Array} props.items Массив с рекомендованными товарами.
 * @param {Function} props.onItemClick Обработчик клика по товару.
 * @param {Function} props.onInViewport Функция, вызываемая при попадании компонента в поле видимости.
 * @param {Function} props.onRequest Функция, инициализирующая загрузку рекомендаций.
 * @param {Function} props.addObserve Функция подписки на Intersection Observer.
 * @param {Function} props.addObserveWithMargin
 * Функция подписки на Intersection Observer с увеличенным пределом попадания в область видимости.
 * @return {ReactElement} Компонент карусели рекомендаций.
 */
export const RecommendationCarousel = ({
  title,
  items,
  onItemClick,
  onInViewport,
  onRequest,
  addObserve,
  addObserveWithMargin,
}) => {
  const sectionRef = useRef();
  const emptyRef = useRef();

  // Отправляем статистку в ОКО, когда компонент попадает в зону видимости
  useEffect(() => {
    isFunction(addObserve)
    && isFunction(onInViewport)
    && sectionRef.current
    && addObserve(sectionRef.current, onInViewport);
  }, [sectionRef.current]);

  // Инициируем загрузку данных, когда компонент почти попал в зону видимости
  useEffect(() => {
    isFunction(addObserveWithMargin)
    && isFunction(onRequest)
    && emptyRef.current
    && addObserveWithMargin(emptyRef.current, onRequest);
  }, [emptyRef.current]);

  return isNonEmptyArray(items)
    ? (
      <section
        className={cx('wrapper')}
        ref={sectionRef}
      >
        <h3 className={cx('title')}>{title}</h3>
        <Carousel
          items={items}
          renderItem={(item, key) => (
            <RecommendedItem
              {...item}
              onItemClick={() => isFunction(onItemClick) && onItemClick(item)}
              key={key}
            />
          )}
          renderControl={CarouselButton}
        />
      </section>
    )
    : <div ref={emptyRef}></div>;
};

RecommendationCarousel.propTypes = {
  /**
   * Заголовок карусели.
   */
  title: Types.string,

  /**
   * Массив элементов карусели.
   */
  items: Types.arrayOf(Types.shape({
    name: Types.string,
    imageSrc: Types.string,
    imageAlt: Types.string,
    url: Types.string,
    price: Types.number,
    currencyGrapheme: Types.string,
    onQuickViewClick: Types.func,
    oldPrice: Types.number,
  })),

  /**
   * Обработчик клика на элементе карусели.
   */
  onItemClick: Types.func,

  /**
   * Функция, будет вызвана при попадании карусели во вьюпорт.
   */
  onInViewport: Types.func,

  /**
   * Функция для получения элементов карусели, будет вызвана при приближении карусели к вьюпорту.
   */
  onRequest: Types.func,

  /**
   * Функция подписки на Intersection Observer, предоставляется HOC'ом.
   */
  addObserve: Types.func,

  /**
   * Функция подписки на Intersection Observer, предоставляется HOC'ом.
   */
  addObserveWithMargin: Types.func,
};

export default withInViewportObserver(
  withInViewportObserver(RecommendationCarousel), {
    rootMargin: '200px 0px 200px 0px',
  }, 'addObserveWithMargin'
);
