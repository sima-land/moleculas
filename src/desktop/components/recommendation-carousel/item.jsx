import React from 'react';
import { Link } from '@dev-dep/ui-nucleons/link';
import { Price } from '@dev-dep/ui-nucleons/price';
import QuickViewSVG from '../../../common/icons/quick-view.svg';
import classnames from 'classnames/bind';
import styles from './item.scss';

const cx = classnames.bind(styles);

/**
 * Компонент элемента карусели рекомендаций.
 * @param {Object} props Свойства компонента.
 * @param {string} props.url Ссылка на товар.
 * @param {string} props.name Название товара.
 * @param {string} props.imageSrc Ссылка на картинку.
 * @param {number} props.price Цена товара.
 * @param {number} props.oldPrice Ссылка на картинку.
 * @param {string} props.currencyGrapheme Графема валюты.
 * @param {Function} props.onClick Обработчик клика по товару.
 * @param {Object} props.quickViewButtonProps Свойства кнопки быстрого просмотра.
 * @return {ReactElement} Элемент.
 */
export const RecommendedItem = ({
  url,
  name,
  imageSrc,
  price,
  oldPrice,
  currencyGrapheme,
  onClick,
  quickViewButtonProps,
  size = {},
}) => (
  <div
    className={cx('root', [
      `size-xs-${size.xs || 4}`,
      `size-s-${size.s || 3}`,
      `size-m-${size.m || 2}`,
      `size-l-${size.l || 2}`,
      `size-xl-${size.xl || 2}`,
    ])}
  >
    {Boolean(imageSrc) && (
      <div className={cx('image-wrapper')}>
        <img
          className={cx('image')}
          alt={name}
          src={imageSrc}
        />
        <div
          {...quickViewButtonProps}
          data-testid='reco-item:quick-view'
          className={cx('quick-view-button')}
          role='button'
        >
          <QuickViewSVG
            className={cx('svg')}
            width={24}
            height={24}
          />
        </div>
      </div>
    )}
    {Boolean(price) && (
      <div className={cx('prices-wrapper')}>
        <Price
          value={price}
          currencyGrapheme={currencyGrapheme}
          className={cx('price')}
        />
        {Boolean(oldPrice) && (
          <Price
            value={oldPrice}
            currencyGrapheme={currencyGrapheme}
            className={cx('old-price')}
            crossedOut
          />
        )}
      </div>
    )}
    <Link
      onClick={onClick}
      className={cx('link')}
      href={url}
      children={name}
      color='black'
    />
  </div>
);
