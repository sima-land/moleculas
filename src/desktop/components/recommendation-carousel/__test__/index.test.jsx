import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { RecommendationCarousel } from '..';
import { items } from '../../../../common/__fixtures__/recommendation-carousel';
import { useMedia } from '../utils';
import { RecommendedItem } from '../item';
import { Link } from '@dev-dep/ui-nucleons/link';

jest.mock('../utils', () => {
  const original = jest.requireActual('../utils');

  const fakeUseMedia = () => Boolean(fakeUseMedia.__flag);

  return {
    ...original,
    __esModule: true,
    useMedia: fakeUseMedia,
  };
});

describe('<RecommendationCarousel />', () => {
  it('should renders correctly', () => {
    const wrapper = mount(
      <RecommendationCarousel
        title='Hello, world!'
        items={items}
        itemSize={{ xs: 3 }}
        className='additional-class'
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle "titleContainer" prop', () => {
    const wrapper = mount(
      <RecommendationCarousel
        title='Hello, world!'
        titleContainer='h1'
        items={items}
        itemSize={{ xs: 3 }}
        className='additional-class'
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render empty ref', () => {
    const wrapper = mount(
      <RecommendationCarousel items={undefined} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should set size depend on media query', () => {
    const wrapper = mount(
      <RecommendationCarousel
        title='Hello, world!'
        items={items}
      />
    );

    expect(wrapper).toMatchSnapshot();

    act(() => {
      useMedia.__flag = true;
      wrapper.setProps({ title: 'Hello!' });
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle "onItemClick" prop', () => {
    const spy = jest.fn();

    const wrapper = mount(
      <RecommendationCarousel
        title='Hello, world!'
        items={items}
        onItemClick={spy}
      />
    );

    expect(spy).toBeCalledTimes(0);

    act(() => {
      wrapper.find(RecommendedItem).at(0).find(Link).simulate('click');
    });
    wrapper.update();

    expect(spy).toBeCalledTimes(1);
  });

  it('should handle "onQuickViewClick" prop', () => {
    const spy = jest.fn();

    const wrapper = mount(
      <RecommendationCarousel
        title='Hello, world!'
        items={items}
        onQuickViewClick={spy}
      />
    );

    expect(spy).toBeCalledTimes(0);

    act(() => {
      wrapper.find(RecommendedItem).at(0).find('[data-testid="reco-item:quick-view"]').simulate('click');
    });
    wrapper.update();

    expect(spy).toBeCalledTimes(1);
  });
});
