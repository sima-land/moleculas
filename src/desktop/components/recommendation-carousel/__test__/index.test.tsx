import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { RecommendationCarousel } from '..';
import { items } from './test-items';
import { useMedia } from '@sima-land/ui-nucleons/hooks/media';
import { HoverCard } from '../hover-card';
import { ProductCard } from '../product-card';
import { ProductInfo } from '../product-info';
import { Link } from '@sima-land/ui-nucleons/link';
import { Carousel } from '@sima-land/ui-nucleons/carousel';
import { Stepper } from '@sima-land/ui-nucleons/stepper';

jest.mock('@sima-land/ui-nucleons/hooks/media', () => {
  const original = jest.requireActual('@sima-land/ui-nucleons/hooks/media');

  const fakeUseMedia = (): boolean => Boolean((fakeUseMedia as any).__flag);

  return {
    ...original,
    __esModule: true,
    useMedia: fakeUseMedia,
  };
});

describe('<RecommendationCarousel />', () => {
  const Find = {
    hoverCard: (wrapper: ReactWrapper) => wrapper.find(HoverCard),
    hoverCardItemName: (wrapper: ReactWrapper) => wrapper.find(HoverCard)
      .find(ProductCard)
      .find(ProductInfo)
      .find(Link)
      .text(),
    item: (wrapper: ReactWrapper) => wrapper.find('[data-testid="reco-item"]'),
  };

  it('should renders correctly', () => {
    const wrapper = mount(
      <RecommendationCarousel
        title='Hello, world!'
        items={items}
        itemSize={{ xs: 3 }}
        className='additional-class'
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle "onItemLinkClick" prop', () => {
    const spy = jest.fn();

    const wrapper = mount(
      <RecommendationCarousel
        items={items}
        onItemLinkClick={spy}
      />
    );

    expect(spy).toBeCalledTimes(0);

    act(() => {
      wrapper.find('a[data-testid="product-info:image-link"]').at(1).simulate('click');
    });
    wrapper.update();

    expect(spy).toBeCalledTimes(1);

    act(() => {
      wrapper.find('a[data-testid="product-info:name-link"]').at(2).simulate('click');
    });
    wrapper.update();

    expect(spy).toBeCalledTimes(2);
  });

  it('should handle "onItemLinkClick" prop missing', () => {
    const wrapper = mount(
      <RecommendationCarousel
        items={items}
      />
    );

    act(() => {
      wrapper.find('a[data-testid="product-info:image-link"]').at(1).simulate('click');
    });
    wrapper.update();

    act(() => {
      wrapper.find('a[data-testid="product-info:name-link"]').at(2).simulate('click');
    });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle "titleTag" prop', () => {
    const wrapper = mount(
      <RecommendationCarousel
        title='Hello, world!'
        titleTag='h1'
        items={items}
        itemSize={{ xs: 3 }}
        className='additional-class'
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render empty ref', () => {
    const wrapper = mount(
      <RecommendationCarousel items={[]} />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should set size depend on media query', () => {
    const wrapper = mount(
      <RecommendationCarousel
        title='Hello, world!'
        items={items}
      />
    );

    expect(wrapper).toMatchSnapshot();

    act(() => {
      (useMedia as any).__flag = true;
      wrapper.setProps({ title: 'Hello!' });
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle item mouseenter event', () => {
    const wrapper = mount(
      <RecommendationCarousel
        title='Hello, world!'
        titleTag='h1'
        items={items}
        itemSize={{ xs: 3 }}
        className='additional-class'
        withHoverCard
      />
    );

    act(() => {
      Find.item(wrapper).at(0).simulate('mouseenter');
    });
    wrapper.update();

    expect(Find.hoverCardItemName(wrapper)).toBe(items[0].product.name);
  });

  it('should handle hover card link clicks', () => {
    const spy = jest.fn();

    const wrapper = mount(
      <RecommendationCarousel
        items={items}
        withHoverCard
        onItemLinkClick={spy}
      />
    );

    act(() => {
      Find.item(wrapper).at(1).simulate('mouseenter');
    });
    wrapper.update();

    expect(Find.hoverCard(wrapper)).toHaveLength(1);

    act(() => {
      Find.hoverCard(wrapper).find('a[data-testid="product-info:image-link"]').simulate('click');
    });
    wrapper.update();

    expect(spy).toBeCalledTimes(1);

    act(() => {
      Find.hoverCard(wrapper).find('a[data-testid="product-info:name-link"]').simulate('click');
    });
    wrapper.update();

    expect(spy).toBeCalledTimes(2);
  });

  it('should handle hover card link clicks without "onItemLinkClick" prop', () => {
    const wrapper = mount(
      <RecommendationCarousel
        items={items}
        withHoverCard
      />
    );

    act(() => {
      Find.item(wrapper).at(1).simulate('mouseenter');
    });
    wrapper.update();

    expect(Find.hoverCard(wrapper)).toHaveLength(1);

    act(() => {
      Find.hoverCard(wrapper).find('a[data-testid="product-info:image-link"]').simulate('click');
    });
    wrapper.update();

    act(() => {
      Find.hoverCard(wrapper).find('a[data-testid="product-info:name-link"]').simulate('click');
    });
    wrapper.update();

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle hover card mouseleave', () => {
    const wrapper = mount(
      <RecommendationCarousel
        title='Hello, world!'
        titleTag='h1'
        items={items}
        itemSize={{ xs: 3 }}
        className='additional-class'
        withHoverCard
      />
    );

    act(() => {
      Find.item(wrapper).at(0).simulate('mouseenter');
    });
    wrapper.update();

    expect(Find.hoverCardItemName(wrapper)).toBe(items[0].product.name);

    act(() => {
      (Find.hoverCard(wrapper).prop('onMouseLeave') as any)();
    });
    wrapper.update();

    expect(Find.hoverCard(wrapper)).toHaveLength(0);
  });

  it('should handle carousel slide', () => {
    jest.useFakeTimers();

    const wrapper = mount(
      <RecommendationCarousel
        title='Hello, world!'
        titleTag='h1'
        items={items}
        itemSize={{ xs: 3 }}
        withHoverCard
      />
    );

    act(() => {
      (wrapper.find(Carousel).prop('onChangeTargetIndex') as any)();
    });
    wrapper.update();

    expect(Find.hoverCard(wrapper)).toHaveLength(0);

    act(() => {
      Find.item(wrapper).at(0).simulate('mouseenter');
    });
    wrapper.update();

    expect(Find.hoverCard(wrapper)).toHaveLength(0);

    jest.advanceTimersByTime(400);

    act(() => {
      Find.item(wrapper).at(0).simulate('mouseenter');
    });
    wrapper.update();

    expect(Find.hoverCardItemName(wrapper)).toBe(items[0].product.name);
  });

  it('should handle cart data/control props', () => {
    const Spy = {
      onItemAdd: jest.fn(),
      onItemChange: jest.fn(),
      onItemSubtract: jest.fn(),
    };

    const wrapper = mount(
      <RecommendationCarousel
        withHoverCard
        items={[
          {
            product: {
              name: 'Test Product',
              imageSrc: 'http://image.com/test',
              url: 'https://www.sima-land.ru/123',
              price: 100,
              currencyGrapheme: '$',
              oldPrice: 200,
            },
            cart: {
              stepText: 'по 1 шт',
              markupText: 'Комплектация + 50$ при покупке до 20 шт',
              qty: 12,
              inCart: true,
            },
          },
        ]}
        onItemAdd={Spy.onItemAdd}
        onItemChange={Spy.onItemChange}
        onItemSubtract={Spy.onItemSubtract}
      />
    );

    act(() => {
      Find.item(wrapper).at(0).simulate('mouseenter');
    });
    wrapper.update();

    expect(wrapper.find(HoverCard).find(Stepper).find('[data-testid="stepper:input"]').prop('value')).toBe('12');

    Object.values(Spy).forEach(fn => {
      expect(fn).toBeCalledTimes(0);
    });

    // add
    act(() => {
      wrapper.find(HoverCard).find(Stepper).find('svg[data-testid="stepper:plus"]').simulate('click');
    });
    wrapper.update();

    expect(Spy.onItemAdd).toBeCalledTimes(1);

    // subtract
    act(() => {
      wrapper.find(HoverCard).find(Stepper).find('svg[data-testid="stepper:minus"]').simulate('click');
    });
    wrapper.update();

    expect(Spy.onItemSubtract).toBeCalledTimes(1);

    // change
    act(() => {
      wrapper.find(HoverCard).find(Stepper).find('[data-testid="stepper:input"]').simulate('change', {
        target: { value: '15' },
      });
    });
    wrapper.update();
    act(() => {
      wrapper.find(HoverCard).find(Stepper).find('[data-testid="stepper:input"]').simulate('blur');
    });
    wrapper.update();

    expect(Spy.onItemChange).toBeCalledTimes(1);
  });
});
