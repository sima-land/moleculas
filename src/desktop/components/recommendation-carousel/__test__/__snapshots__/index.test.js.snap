// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecommendationCarousel /> should match a snapshot with items 1`] = `
<section
  className="wrapper"
>
  <h3
    className="title"
  >
    Test title
  </h3>
  <withGlobalListeners(Carousel)
    items={
      Array [
        Object {
          "currencyGrapheme": "₽",
          "imageAlt": "Ножницы портновские, 19 см, цвет чёрный, Ножницы",
          "imageSrc": "https://cdn3.static1-sima-land.com/items/3572134/0/280.jpg?v=1594121833",
          "name": "Ножницы портновские, 19 см, цвет чёрный",
          "oldPrice": null,
          "price": 261,
          "url": "",
        },
        Object {
          "currencyGrapheme": "₽",
          "imageAlt": "Ножницы портновские, 31 см, цвет чёрный/голубой, Ножницы",
          "imageSrc": "https://cdn3.static1-sima-land.com/items/2300382/0/280.jpg?v=1594121798",
          "name": "Ножницы портновские, 31 см, цвет чёрный/голубой",
          "oldPrice": null,
          "price": 322,
          "url": "",
        },
        Object {
          "currencyGrapheme": "₽",
          "imageAlt": "Ножницы портновские, 21,5 см, цвет чёрный/голубой, Ножницы",
          "imageSrc": "https://cdn3.static1-sima-land.com/items/2300383/0/280.jpg?v=1594121773",
          "name": "Ножницы портновские, 21,5 см, цвет чёрный/голубой",
          "oldPrice": null,
          "price": 238,
          "url": "",
        },
        Object {
          "currencyGrapheme": "₽",
          "imageAlt": "Ножницы портновские, 23 см, цвет МИКС, Ножницы",
          "imageSrc": "https://cdn3.static1-sima-land.com/items/1550239/0/280.jpg?v=1600179085",
          "name": "Ножницы портновские, 23 см, цвет МИКС",
          "oldPrice": null,
          "price": 171,
          "url": "",
        },
        Object {
          "currencyGrapheme": "₽",
          "imageAlt": "Ножницы портновские, одинаковые кольца, 23 см, цвет чёрный, Ножницы",
          "imageSrc": "https://cdn3.static1-sima-land.com/items/4002852/0/280.jpg?v=1594121894",
          "name": "Ножницы портновские, одинаковые кольца, 23 см, цвет чёрный",
          "oldPrice": null,
          "price": 483,
          "url": "",
        },
        Object {
          "currencyGrapheme": "₽",
          "imageAlt": "Ножницы портновские, 22,5 см, цвет чёрный, Ножницы",
          "imageSrc": "https://cdn3.static1-sima-land.com/items/4005918/0/280.jpg?v=1594121898",
          "name": "Ножницы портновские, 22,5 см, цвет чёрный",
          "oldPrice": null,
          "price": 988,
          "url": "",
        },
        Object {
          "currencyGrapheme": "₽",
          "imageAlt": "Ножницы портновские, 22 см, цвет серебряный, Ножницы",
          "imageSrc": "https://cdn3.static1-sima-land.com/items/2300389/0/280.jpg?v=1594121830",
          "name": "Ножницы портновские, 22 см, цвет серебряный",
          "oldPrice": null,
          "price": 619,
          "url": "",
        },
        Object {
          "currencyGrapheme": "₽",
          "imageAlt": "Ножницы портновские, с увеличенным углом наклона, 27 см, цвет чёрный, Ножницы",
          "imageSrc": "https://cdn3.static1-sima-land.com/items/2117838/0/280.jpg?v=1594121766",
          "name": "Ножницы портновские, с увеличенным углом наклона, 27 см, цвет чёрный",
          "oldPrice": 595,
          "price": 539,
          "url": "",
        },
        Object {
          "currencyGrapheme": "₽",
          "imageAlt": "Ножницы портновские, с прорезиненной ручкой, 20 см, цвет чёрный/серый, Ножницы",
          "imageSrc": "https://cdn3.static1-sima-land.com/items/1108673/0/280.jpg?v=1594103956",
          "name": "Ножницы портновские, с прорезиненной ручкой, 20 см, цвет чёрный/серый",
          "oldPrice": 717,
          "price": 599,
          "url": "",
        },
        Object {
          "currencyGrapheme": "₽",
          "imageAlt": "Ножницы портновские, скошенное лезвие, 15 см, цвет серебряный, Ножницы",
          "imageSrc": "https://cdn3.static1-sima-land.com/items/4150936/0/280.jpg?v=1580193932",
          "name": "Ножницы портновские, скошенное лезвие, 15 см, цвет серебряный",
          "oldPrice": null,
          "price": 253,
          "url": "",
        },
      ]
    }
    renderControl={[Function]}
    renderItem={[Function]}
  />
  <div
    className="tooltip"
  >
    Быстрый просмотр
  </div>
</section>
`;

exports[`<RecommendationCarousel /> should match a snapshot without items 1`] = `<div />`;
